include/deal.II/multigrid/mg_constrained_dofs.h

```diff
@@ -146,13 +168,53 @@ MGConstrainedDoFs::initialize(const DoFHandler<dim,spacedim> &dof)
 {
   boundary_indices.clear();
   refinement_edge_indices.clear();
+  level_constraints.clear();
 
   const unsigned int nlevels = dof.get_triangulation().n_global_levels();
 
-  //At this point refinement_edge_indices is empty.
+  //At this point level_constraint and refinement_edge_indices are empty.
+  level_constraints.resize(nlevels);
   refinement_edge_indices.resize(nlevels);
   for (unsigned int l=0; l<nlevels; ++l)
-    refinement_edge_indices[l] = IndexSet(dof.n_dofs(l));
+    {
+      IndexSet relevant_dofs;
+      DoFTools::extract_locally_relevant_level_dofs(dof,l,relevant_dofs);
+      level_constraints[l].reinit(relevant_dofs);
```

Comment:
We test the ghosts with periodicity in `mpi/mg_ghost_dofs_periodic_01.cc`. It might happen (I don't remember exactly how we implemented it, I remember that we needed to have a ghost layer that is rather thick) that a ghosted cell sits at the periodic boundary. Since we go through the periodic neighbors in that case, we might hit a cell with invalid subdomain id. @tcclevenger can you add a check `cell->periodic_neighbor(f)->level_subdomain_id() == numbers::artificial_subdomain_id` and skip the part starting from line 201 in that case?

With that check, I think we should be fine. Whenever a cell is ghosted but a parent cell is a locally owned level cell (which is needed in the transfer), we also have the cell behind the periodic face on that ghosted cell among the ghosts according to the way we construct the periodicity.